### PR TITLE
Fix leaks in DPS enrollment tests

### PR DESCRIPTION
**Pull request by ngastelum-ms**

### CODE REVIEW

File: e2e/test/provisioning/ProvisioningE2ETests.cs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 using Microsoft.Azure.Devices.Provisioning.Service;
 using Microsoft.Azure.Devices.Shared;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Tpm2Lib;
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningE2ETests.cs:22]: Why was this needed?

Comment (ngastelum-ms) [reply e2e/test/provisioning/ProvisioningE2ETests.cs]: Accidental autocomplete, removed

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 2]
```diff
@@ -153,101 +153,135 @@ public async Task ProvisioningServiceClient_GetIndividualEnrollmentAttestation(A
...
+                await RetryOperationHelper
+                                .RetryOperationsAsync(
+                                    async () =>
+                                    {
+                                        attestationMechanism = await provisioningServiceClient.GetIndividualEnrollmentAttestationAsync(individualEnrollment.RegistrationId);
+                                    },
+                                    s_provisioningServiceRetryPolicy,
+                                    s_retryableExceptions,
+                                    Logger)
+                                .ConfigureAwait(false);
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:183]:
```suggestion
                await RetryOperationHelper
                    .RetryOperationsAsync(
                        async () =>
                        {
                            attestationMechanism = await provisioningServiceClient.GetIndividualEnrollmentAttestationAsync(individualEnrollment.RegistrationId);
                        },
                        s_provisioningServiceRetryPolicy,
                        s_retryableExceptions,
                        Logger)
                    .ConfigureAwait(false);
```

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 3]
```diff
@@ -153,101 +153,135 @@ public async Task ProvisioningServiceClient_GetIndividualEnrollmentAttestation(A
...
+                await RetryOperationHelper
+                                .RetryOperationsAsync(
+                                    async () =>
+                                    {
+                                        attestationMechanism = await provisioningServiceClient.GetEnrollmentGroupAttestationAsync(enrollmentGroup.EnrollmentGroupId);
+                                    },
+                                    s_provisioningServiceRetryPolicy,
+                                    s_retryableExceptions,
+                                    Logger)
+                                .ConfigureAwait(false);
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:254]: Tabbed in too far

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 4]
```diff
@@ -287,53 +321,63 @@ public async Task ProvisioningServiceClient_IndividualEnrollments_Create_Ok(stri
...
+                        reprovisionPolicy,
+                        allocationPolicy,
+                        customAllocationDefinition,
+                        iotHubsToProvisionTo,
+                        null,
+                        Logger).ConfigureAwait(false);
+
+                    IndividualEnrollment individualEnrollmentResult = null;
+                    await RetryOperationHelper
+                                    .RetryOperationsAsync(
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:342]: again

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 5]
```diff
@@ -353,51 +397,61 @@ public async Task ProvisioningServiceClient_GroupEnrollments_Create_Ok(string pr
...
+                        reprovisionPolicy,
+                        allocationPolicy,
+                        customAllocationDefinition,
+                        iothubs,
+                        null,
+                        Logger).ConfigureAwait(false);
+
+                    EnrollmentGroup enrollmentGroupResult = null;
+                    await RetryOperationHelper
+                                    .RetryOperationsAsync(
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:417]: again

File: e2e/test/provisioning/ReprovisioningE2ETests.cs [hunk 6]
```diff
@@ -327,63 +327,82 @@ private async Task ProvisioningDeviceClient_ReprovisioningFlow_DoNotReprovision(
...
-            else
+            finally
             {
-                Logger.Trace($"Deleting test enrollment type {attestationType}-{enrollmentType} with registration Id {security.GetRegistrationID()}.");
-                await DeleteCreatedEnrollmentAsync(enrollmentType, security, groupId, Logger).ConfigureAwait(false);
-            }
+                try
+                {
+                    if (
+                        auth == null
```
Comment (drwill-ms) [inline e2e/test/provisioning/ReprovisioningE2ETests.cs:378]:
```suggestion
                    if (auth == null
```

**Review by drwill-ms — APPROVED** [review]

File: provisioning/service/samples/How To/GroupCertificateVerificationSample/GroupCertificateVerificationSample.csproj [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
-    <OutputType>Exe</OutputType>
-    <TargetFramework>net462</TargetFramework>
+	  <TargetFramework>net6.0</TargetFramework>
+	  <LangVersion>9.0</LangVersion>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="BouncyCastle" Version="1.8.9" />
+    <PackageReference Include="BouncyCastle" Version="1.8.9">
+	  <NoWarn>NU1701</NoWarn>
```
Comment (abhipsaMisra) [inline provisioning/service/samples/How To/GroupCertificateVerificationSample/GroupCertificateVerificationSample.csproj:10]: why is this now required?

File: e2e/test/provisioning/ProvisioningE2ETests.cs [hunk 8]
```diff
@@ -1228,6 +1188,49 @@ public static bool ImplementsWebProxy(Client.TransportType transportProtocol)
             throw new NotSupportedException($"Unknown transport: '{transportProtocol}'.");
```
Comment (abhipsaMisra) [inline e2e/test/provisioning/ProvisioningE2ETests.cs:1188]: Unrelated to this PR, but I thought it was cool so I thought I'd share:
Intellisense recommends the following for `ImplementsWebProxy`

```c#
return transportProtocol switch
    {
        Client.TransportType.Mqtt_WebSocket_Only or Client.TransportType.Amqp_WebSocket_Only => true,
        _ => false,
    };
```

File: e2e/test/provisioning/ProvisioningE2ETests.cs [hunk 9]
```diff
@@ -1228,6 +1188,49 @@ public static bool ImplementsWebProxy(Client.TransportType transportProtocol)
...
         }
 
+        private async Task CleanupEnrollmentsAsync(AttestationMechanismType attestationMechanismType, EnrollmentType enrollmentType, string registrationId, string groupId)
```
Comment (abhipsaMisra) [inline e2e/test/provisioning/ProvisioningE2ETests.cs:1191]: These don't have a reference to the instance logger now, so they can be static

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 10]
```diff
@@ -149,98 +149,132 @@ public async Task ProvisioningServiceClient_GetIndividualEnrollmentAttestation(A
...
+            finally
             {
-                attestationMechanism.Type.Should().Be(AttestationMechanismType.Tpm);
-
-                var tpmAttestation = (TpmAttestation)attestationMechanism.GetAttestation();
-                tpmAttestation.EndorsementKey.Should().Be(((TpmAttestation)individualEnrollment.Attestation).EndorsementKey);
-                tpmAttestation.StorageRootKey.Should().Be(((TpmAttestation)individualEnrollment.Attestation).StorageRootKey);
+                if (individualEnrollment != null)
+                {
+                    await DeleteCreatedEnrollmentAsync(EnrollmentType.Individual, individualEnrollment.RegistrationId, null);
```
Comment (abhipsaMisra) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:213]: we'll also want to dispose any certificates associated with the x509 attestation mechanism

Comment (abhipsaMisra) [reply e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:213]: same for the group enrollment as well

Comment (abhipsaMisra) [reply e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:213]: The provisioning service E2E tests don't seem to be creating an x509 certificate based enrollments, so you can either add those or add a comment calling out that we're not testing that in here.
Iirc we cover x509 certificate based scenarios pretty extensively in the provisioning device client E2E tests so I'd be fine with us simply adding a comment here calling out that we're not testing x509 here.

**Review by abhipsaMisra — APPROVED** [review]
Check the ProvisioningServiceE2ETests to confirm that we're not testing x509 based enrollments in there, and then remove all x509 code from the helper methods in here.
We do test x509 based scenarios pretty extensively in the provisioning device client E2E tests, so I'd be fine with us simply adding a comment here calling out that we're not testing x509 here.

File: e2e/test/provisioning/ProvisioningE2ETests.cs (hunk 8 above)
Comment (ngastelum-ms) [reply e2e/test/provisioning/ProvisioningE2ETests.cs:1188]: Oh wow that is cool

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs (hunk 10 above)
Comment (ngastelum-ms) [reply e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:206]: Added comment and removed unused x509 code
